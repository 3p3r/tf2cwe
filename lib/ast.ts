type Value = string | number | boolean | Reference;
export type PropertyValue = Value | Value[] | Record<string, Value | Value[]>;

export type Attribute = string | number;

export enum Type {
  Resource,
  Output,
  Variable,
  Reference,
  Config,
}

export interface Node {
  readonly id: number;
  is: (type: Type) => boolean;
}

export interface Config extends Node {}

// todo: remove Node from name
export interface Reference extends Node {
  readonly target: string | number;
  readonly property: Attribute[];
}

export interface Output extends Node {
  readonly name: string;
  readonly type: string;
  readonly value: string | Reference;
}

export interface Resource extends Node {
  readonly name: string;
  readonly type: string;
  readonly properties: Record<string, PropertyValue>;
}

export interface Variable extends Node {
  readonly name: string;
  readonly type: string;
  readonly default?: string;
  readonly description?: string;
  // TODO: validation should be implemented
  // readonly validation?: string;
  readonly sensitive?: boolean;
  readonly nullable?: boolean;
}
