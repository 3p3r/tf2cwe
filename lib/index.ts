export * as input from "./in";
export * as output from "./out";
export * as ast from "./ast";
export * as hooks from "./hooks";
