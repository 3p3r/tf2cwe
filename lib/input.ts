import fs from "fs/promises";
import path from "path";

interface WalkOptions {
  depth?: number;
  extensions?: string[];
}

type entries = { [key: string]: string[] };

function supportedExtension(list: string[], ext: string) {
  return list[0] === "*" || list.includes(ext);
}

async function run(folder: string, depth = -1, opts: Required<WalkOptions>): Promise<entries> {
  depth++;
  let inputs: entries = {};
  const dirents = await fs.readdir(folder, { withFileTypes: true });
  for (const stat of dirents) {
    const child = path.resolve(folder, stat.name);
    if (stat.isDirectory()) {
      if (depth > opts.depth) {
        continue;
      }
      const extGroups = await run(child, depth, opts);
      for (const [ext, arr] of Object.entries(extGroups)) {
        !inputs[ext] && (inputs[ext] = []);
        inputs[ext] = inputs[ext].concat(arr);
      }
      continue;
    }
    const ext = path.extname(stat.name);
    if (!supportedExtension(opts.extensions, ext)) {
      continue;
    }
    !inputs[ext] && (inputs[ext] = []);
    inputs[ext].push(child);
  }
  return inputs;
}

export async function walk(input: string, opts: WalkOptions = {}): Promise<entries> {
  const options = {
    depth: Infinity,
    extensions: ["*"],
    ...opts,
  };
  const abs = path.resolve(input);
  const dirent = await fs.stat(abs);
  if (!dirent.isDirectory()) {
    const ext = path.extname(abs);
    if (!supportedExtension(options.extensions, ext)) {
      throw new Error(`input must be a directory or file of type ${options.extensions.join(", ")}`);
    }
    return { [ext]: [abs] };
  }
  return run(abs, options.depth, options);
}
