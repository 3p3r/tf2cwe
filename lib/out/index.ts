import * as cwe from "./cwe";
import { ast } from "../";

export * as cwe from "./cwe";
export enum Language {
  CloudTrail = "cwe",
}

interface CompileOptions {
  input: ast.Node[];
  language: Language;
}

export async function compile(options: CompileOptions & { language: Language.CloudTrail }): Promise<cwe.types.Event[]>;
export async function compile(options: CompileOptions) {
  if (!options.input) {
    throw new Error("tf2cwe compile requires an input");
  }
  switch (options.language) {
    case Language.CloudTrail:
      return await cwe.compile(options.input);
    default:
      break;
  }
}
