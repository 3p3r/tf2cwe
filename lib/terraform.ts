import assert from "assert";
import fs from "fs";
import Parser from "web-tree-sitter";
import * as adk from "./vendor/adk";

enum NodeType {
  provider = "provider",
  resource = "resource",
  module = "module",
  output = "output",
  variable = "variable",
}

type Block = Resource;

interface Property {}

interface Resource {
  node: Parser.SyntaxNode;
  properties: Property[];
}

interface Context {
  readonly resources: Resource[];
  readonly parser: Parser;
  readonly blockCache: Map<number, Block>;
  readonly blockRoots: Set<number>;
  readonly node: Parser.SyntaxNode;
}

function getAllRuleNodeTypes(): string[] {
  return Object.values(NodeType);
}

function emitProperty(key: Parser.SyntaxNode, value: Parser.SyntaxNode, context: Context) {
  console.log(key, value, context);
  return { [key.text]: value };
  // (template_expr (quoted_template (template_literal) @lit (template_interpolation) @temp ) )
  // (template_expr (quoted_template (template_literal)? @lit (template_interpolation)? @int  ))

  // gets parts of a literal
  // (template_expr (quoted_template (template_literal) @lit  ) )

  // string value
  // context.parser.getLanguage().query("(literal_value (string_lit (template_literal) @str))").captures(value);

  //
}

function emitBlockResource(context: Context): Resource {
  const body = context.parser
    .getLanguage()
    .query("(attribute (identifier) @key (expression) @value)")
    .captures(context.node.parent);
  assert.ok(body.length % 2 === 0, "missing pairs for properties in resource block");
  const resource: Resource = {
    node: context.node,
    properties: [],
  };
  for (let i = 0; i < body.length; i += 2) {
    resource.properties.push(emitProperty(body[i].node, body[i + 1].node, context));
  }
  return resource;
}

function emitBlock(context: Context): Block {
  if (context.blockCache.has(context.node.id)) {
    return context.blockCache.get(context.node.id);
  }
  let iterator: Parser.SyntaxNode | null = context.node.parent;
  let root = true;
  while (iterator) {
    if (getAllRuleNodeTypes().includes(iterator.text)) {
      root = false;
      break;
    }
    iterator = iterator.parent;
  }
  const block = (() => {
    switch (context.node.text) {
      case NodeType.resource:
        return emitBlockResource(context);
      case NodeType.provider:
      case NodeType.module:
      case NodeType.variable:
      case NodeType.output:
        return undefined;
      default:
        assert.ok(false, `unhandled node type: ${context.node.type}`);
    }
  })();
  if (block) {
    if (root) {
      context.blockRoots.add(block.node.id);
    }
    context.blockCache.set(context.node.id, block);
  }
  return block;
}

export async function compile(parser: Parser, sources: { [key: string]: string[] }) {
  if (!sources[".tf"]) {
    throw new Error("Input contains no .tf files. only .tf files are supported");
  }
  const blockCache = new Map<number, Block>();
  const blockRoots = new Set<number>();

  const gigaTf = sources[".tf"].reduce((prev, next) => {
    return prev + fs.readFileSync(next, { encoding: "utf-8" });
  }, "");

  const tree = parser.parse(gigaTf);
  const blocks = parser
    .getLanguage()
    .query("(block (identifier) @block (string_lit (template_literal)))")
    .captures(tree.rootNode);

  for (const block of blocks) {
    emitBlock({ node: block.node, blockCache, blockRoots, parser, resources: [] });
  }

  // console.log(tree.rootNode.toString());
  // const bodies: string[] = [];
  // const regoRulenames: string[] = [];
  // const kebabRegExp = new RegExp(/[^a-zA-Z0-9]|-{1,}/, "g");

  // const blockRoots = new Set<string>();
  // const token = "cap";
  // const ruleQueries = getAllRuleNodeTypes()
  //   .map((q) => `(${q}) @${token}`)
  //   .map((q) => parser.getLanguage().query(q))
  //   .map((q) => q.captures(tree.rootNode))
  //   .flatMap((q) => q);
  // const rules: string[] = [];
  // for (const ruleQuery of ruleQueries) {
  //   assert.ok(ruleQuery.name === token);
  //   const node = ruleQuery.node;
  //   emitRule({ node, rules, blockCache, blockRoots });
  // }
  // let header = "";
  // if (sources.length === 1) {
  //   header = "package rule2rego\ndefault allow := false";
  // } else {
  //   const pkgPath = path.relative(sourceRoot, sourcePath).slice(0, -5);
  //   const ruleName = pkgPath.replace(kebabRegExp, "");
  //   regoRulenames.push(ruleName);
  //   header = `package rule2rego.${ruleName}\ndefault allow := false`;
  // }
  // const init = [...new Set(Array.from(blockCache.values()))].map((i) => `default ${i} := false`).join("\n");
  // const policy = `\n${rules.join("\n")}`;
  // const footer = `allow {\n\t${Array.from(blockRoots).join("\n\t")}\n}`;
  // const body = `${header}\n${init}${policy}\n${footer}`.replace(/\n\n/g, "\n");
  // if (options.input === process.argv[2]) {
  //   console.log(body);
  //   console.log("");
  // }
  // bodies.push(body);

  // if (sources.length > 1) {
  //   // add the main
  //   const body = `package rule2rego\ndefault allow := false\nallow {\n  data.rule2rego.${regoRulenames.join(
  //     ".allow\n}\nallow {\n  data.rule2rego."
  //   )}.allow\n}`;
  //   bodies.splice(0, 0, body);
  //   if (options.input === process.argv[2]) {
  //     console.log(body);
  //     console.log("");
  //   }
  // }
  // return bodies;
}
