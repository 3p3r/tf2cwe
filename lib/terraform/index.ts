import assert from "assert";
import fs from "fs";
import Parser from "web-tree-sitter";
// import * as adk from "./vendor/adk";
import * as ast from "../ast";
import { Context } from "./types";

import { emitBlockResource } from "./resource";
import { emitBlockVariable } from "./variable";

enum BlockType {
  Resource = "resource",
  Provider = "provider",
  Module = "module",
  Output = "output",
  Variable = "variable",
  Local = "locals",
}

function getAllBlockTypes(): string[] {
  return Object.values(BlockType);
}

function emitBlock(context: Context, node: Parser.SyntaxNode): ast.Node {
  // if (context.blockCache.has(node.id)) {
  //   return context.blockCache.get(node.id);
  // }
  assert.ok(node.type === "block", `received node type ${node.type}. expected block`);
  let iterator: Parser.SyntaxNode | null = node.parent;
  let root = true;
  while (iterator) {
    if (getAllBlockTypes().includes(iterator.text)) {
      root = false;
      break;
    }
    iterator = iterator.parent;
  }
  const block = (() => {
    switch (node.firstNamedChild.text) {
      case BlockType.Resource:
        const resource = emitBlockResource(context, node);
        context.blocks.push(resource);
        return resource;
      case BlockType.Variable:
        const variable = emitBlockVariable(context, node);
        context.blocks.push(variable);
        return variable;
      case BlockType.Provider:
      case BlockType.Module:
      case BlockType.Output:
      case BlockType.Local:
        return undefined;
      default:
        assert.ok(false, `unhandled node type: ${node.type}`);
    }
  })();
  // if (block) {
  //   if (root) {
  //     context.blockRoots.add(block.id);
  //   }
  //   context.blockCache.set(node.id, block);
  // }
  return block;
}

export async function compile(parser: Parser, sources: { [key: string]: string[] }) {
  if (!sources[".tf"]) {
    throw new Error("Input contains no .tf files. only .tf files are supported");
  }
  // const blockCache = new Map<number, Block>();
  // const blockRoots = new Set<number>();

  const gigaTf = sources[".tf"].reduce((prev, next) => {
    return prev + fs.readFileSync(next, { encoding: "utf-8" });
  }, "");

  const tree = parser.parse(gigaTf);
  const blocks = parser
    .getLanguage()
    .query("(block) @block")
    .captures(tree.rootNode)
    .flatMap((q) => q);
  // getAllBlockTypes()
  //   .map((q) => `(${q}) @block`)
  //   .map((q) => parser.getLanguage().query(q))
  //   .map((q) => q.captures(tree.rootNode))
  //   .flatMap((q) => q);
  const context: Context = { node: undefined /*, blockCache, blockRoots*/, parser, blocks: [] };
  for (const block of blocks) {
    emitBlock({ ...context, node: block.node }, block.node);
  }
  console.log(JSON.stringify(context.blocks, undefined, 2));
}
