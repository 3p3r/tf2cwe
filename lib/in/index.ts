import * as terraform from "./terraform";
import type { Node } from "../ast";

export * as ast from "../ast";
export enum Language {
  Terraform = "hcl",
}

interface CompileOptions {
  input: string;
  language: Language;
}

export async function compile(options: CompileOptions): Promise<Node[]> {
  if (!options.input) {
    throw new Error("tf2cwe compile requires an input");
  }

  switch (options.language) {
    case Language.Terraform:
      return await terraform.compile(options.input);
    default:
      throw new Error(`unhandled language: ${options.input}`);
      break;
  }
}
