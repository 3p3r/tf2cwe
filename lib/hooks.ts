import { AsyncSeriesBailHook, AsyncSeriesWaterfallHook } from "tapable";
import { ast } from "./in";
import { types } from "./out/cwe";

export const out = {
  cwe: {
    /**
     * Called after a resource has been created but before it is emitted to the final output. Return any
     * value to remove the resource from the final output
     */
    excludeResource: new AsyncSeriesBailHook<[types.Event, ast.Resource, string], undefined | any>([
      "event",
      "resource",
      "action",
    ]),
    modifyResource: new AsyncSeriesWaterfallHook<[types.Event, ast.Resource, string]>(["event", "resource", "action"]),
    finalize: new AsyncSeriesWaterfallHook<[types.Event[]]>(["events"]),
  },
};
