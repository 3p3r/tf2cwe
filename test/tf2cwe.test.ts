import assert from "assert";
import { ast, output } from "../lib";
import { fixtures } from "./common";

async function getTerraformFixture() {
  return (await fixtures()).filter((f) => f.name === "terraform")[0];
}
describe("terraform HCL to CWE", () => {
  it("should pass", async () => {
    const fixture = await getTerraformFixture();
    const out = output.compile({ language: output.Language.CloudTrail, input: fixture.nodes });
    assert.ok(out);
  });
});
