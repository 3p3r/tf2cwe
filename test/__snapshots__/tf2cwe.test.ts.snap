// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terraform HCL to CWE should pass 1`] = `
{
  "account": "123456789012",
  "detail": {
    "awsRegion": "us-east-1",
    "eventName": "CreateBucket",
    "eventSource": "ec2.amazonaws.com",
    "eventTime": Any<String>,
    "eventType": 0,
    "eventVersion": "1.08",
    "requestID": "9D767BCC3B4E7487",
    "requestParameters": {},
    "responseElements": null,
    "sourceIPAddress": "100.100.100.100",
    "userAgent": "signin.amazonaws.com",
    "userIdentity": {
      "accountId": "123456789012",
      "arn": "arn:aws:iam::123456789012:root",
      "principalId": "123456789012",
      "sessionContext": {
        "attributes": {
          "creationDate": Any<String>,
          "mfaAuthenticated": "false",
        },
      },
      "type": 0,
    },
  },
  "detail-type": "AWS API Call via CloudTrail",
  "region": "us-east-1",
  "resources": [
    "arn:aws:ec2:us-east-1:123456789012:instance/app_server2",
  ],
  "source": "aws.ec2",
  "time": Any<String>,
  "version": "0",
}
`;
